## AOC 2024/23
### Part 1

--- Day 23: LAN Party ---

As The Historians wander around a secure area at Easter Bunny HQ, you come across posters for a LAN party scheduled for today! Maybe you can find it; you connect to a nearby datalink port and download a map of the local network (your puzzle input).

The network map provides a list of every connection between two computers. For example:

```
kh-tc
qp-kh
de-cg
ka-co
yn-aq
qp-ub
cg-tb
vc-aq
tb-ka
wh-tc
yn-cg
kh-ub
ta-co
de-co
tc-td
tb-wq
wh-td
ta-ka
td-qp
aq-cg
wq-ub
ub-vc
de-ta
wq-aq
wq-vc
wh-yn
ka-de
kh-ta
co-tc
wh-qp
tb-vc
td-yn
```

Each line of text in the network map represents a single connection; the line kh-tc represents a connection between the computer named kh and the computer named tc. Connections aren't directional; tc-kh would mean exactly the same thing.

LAN parties typically involve multiplayer games, so maybe you can locate it by finding groups of connected computers. Start by looking for sets of three computers where each computer in the set is connected to the other two computers.

In this example, there are 12 such sets of three inter-connected computers:

```
aq,cg,yn
aq,vc,wq
co,de,ka
co,de,ta
co,ka,ta
de,ka,ta
kh,qp,ub
qp,td,wh
tb,vc,wq
tc,td,wh
td,wh,yn
ub,vc,wq
```

If the Chief Historian is here, and he's at the LAN party, it would be best to know that right away. You're pretty sure his computer's name starts with t, so consider only sets of three computers where at least one computer's name starts with t. That narrows the list down to 7 sets of three inter-connected computers:

```
co,de,ta
co,ka,ta
de,ka,ta
qp,td,wh
tb,vc,wq
tc,td,wh
td,wh,yn
```

Find all the sets of three inter-connected computers. How many contain at least one computer with a name that starts with t?

To begin, get your puzzle input.

**AI Prompt**

I'm a skilled python developer, please help go consider various approaches for solving this problem and their pros/cons. I prefer simpler and understandable solutions, easily readable also for other developers. It's ok to generate code with testcases without explanation. I'll ask if I do not understand it. Always wrap test code into a function, to not clutter the global variable namespace.

**AI Response**

For this graph problem, we have a few approaches to consider:

1. Using NetworkX library:
   - Pros: Built-in functions for finding cliques of size 3
   - Cons: Might be overkill for this simple problem, adds dependency
   
2. Simple set-based approach:
   - Pros: Very readable, straightforward implementation
   - Cons: Might be slightly less efficient for very large graphs
   
3. Adjacency matrix/list approach:
   - Pros: Could be more efficient for larger graphs
   - Cons: More complex implementation

Would you like me to implement a solution using any of these approaches? I'd personally lean towards the set-based approach since it's simple and the input size seems manageable.

Here's a test case implementation to get us started:

```python
def test_find_triangles():
    test_input = """
kh-tc
qp-kh
de-cg
ka-co
yn-aq
qp-ub
cg-tb
vc-aq
tb-ka
wh-tc
yn-cg
kh-ub
ta-co
de-co
tc-td
tb-wq
wh-td
ta-ka
td-qp
aq-cg
wq-ub
ub-vc
de-ta
wq-aq
wq-vc
wh-yn
ka-de
kh-ta
co-tc
wh-qp
tb-vc
td-yn
""".strip()
    
    assert solve_part1(test_input) == 7

```

Which approach would you prefer to explore first?

**AI Prompt**

load full data

**AI Response**

I'll help you load the puzzle input using the `advent-of-code-data` library which is available in the environment:

```python
from aocd import get_data
data = get_data(day=23, year=2024)
print(f"Data length: {len(data.splitlines())} lines")
print("\nFirst few lines:")
print('\n'.join(data.splitlines()[:5]))
```

Would you like to proceed with implementing one of the approaches after we see the actual data size?

In [ ]:
def parse_input(s: str) -> list[tuple[str, str]]:
    return [edge.split('-') for edge in s.split('\n')]

sample_input = parse_input("""
kh-tc
qp-kh
de-cg
ka-co
yn-aq
qp-ub
cg-tb
vc-aq
tb-ka
wh-tc
yn-cg
kh-ub
ta-co
de-co
tc-td
tb-wq
wh-td
ta-ka
td-qp
aq-cg
wq-ub
ub-vc
de-ta
wq-aq
wq-vc
wh-yn
ka-de
kh-ta
co-tc
wh-qp
tb-vc
td-yn
""".strip()
)
sample_input

[['kh', 'tc'],
 ['qp', 'kh'],
 ['de', 'cg'],
 ['ka', 'co'],
 ['yn', 'aq'],
 ['qp', 'ub'],
 ['cg', 'tb'],
 ['vc', 'aq'],
 ['tb', 'ka'],
 ['wh', 'tc'],
 ['yn', 'cg'],
 ['kh', 'ub'],
 ['ta', 'co'],
 ['de', 'co'],
 ['tc', 'td'],
 ['tb', 'wq'],
 ['wh', 'td'],
 ['ta', 'ka'],
 ['td', 'qp'],
 ['aq', 'cg'],
 ['wq', 'ub'],
 ['ub', 'vc'],
 ['de', 'ta'],
 ['wq', 'aq'],
 ['wq', 'vc'],
 ['wh', 'yn'],
 ['ka', 'de'],
 ['kh', 'ta'],
 ['co', 'tc'],
 ['wh', 'qp'],
 ['tb', 'vc'],
 ['td', 'yn']]

In [ ]:
full_input = parse_input(get_data(day=23, year=2024))

print(f"nodes: {len(set([e[0] for e in full_input] + [e[1] for e in full_input]))}, edges: {len(full_input)}")

nodes: 520, edges: 3380


**AI Prompt**

estimate complexity of finding cliques of size 3 in graph with 520 nodes and 3380 edges

**AI Response**

Let's analyze the complexity for different approaches:

1. Brute force (checking all possible triplets of nodes):
   - O(n³) where n is number of nodes
   - With 520 nodes: about 140 million combinations (520 choose 3)
   - Each combination needs edge checking

2. Edge-based approach (for each edge, find common neighbors):
   - O(m * d) where m is number of edges and d is max degree
   - With 3380 edges and avg degree ~13 (3380*2/520)
   - Much better: roughly 3380 * 13 = ~44k operations

3. NetworkX's algorithm:
   - Uses edge-based approach internally
   - Similar complexity to approach #2

The edge-based approach would be significantly more efficient than brute force here. Would you like me to elaborate on any of these approaches or proceed with implementation?

**AI Prompt**

and for the Adjacency matrix/list approach ?

**AI Response**

For the Adjacency matrix/list approaches:

Adjacency Matrix:
- Space: O(n²) = 520² ≈ 270k elements
- Finding triangles: O(n³) operations checking matrix cells
- Not really beneficial here due to sparse graph (3380 edges vs possible 135k edges)

Adjacency List:
- Space: O(m) = 3380 edges
- Finding triangles: O(m * d) similar to edge-based approach
- For each edge (u,v), check intersection of neighbor lists of u and v
- With good hash-based sets, intersection is O(min(d₁,d₂))

The Adjacency List approach is essentially equivalent to the edge-based approach in complexity. Would you like to proceed with implementing one of these approaches?

**AI Prompt**

lets implement the edge-base approach

**AI Response**

I'll help you implement the edge-based approach. Here's the basic structure:

```python
from collections import defaultdict

def build_graph(edges):
    # Create adjacency list using defaultdict
    graph = defaultdict(set)
    for u, v in edges:
        graph[u].add(v)
        graph[v].add(u)
    return graph

def find_triangles(edges):
    graph = build_graph(edges)
    triangles = set()
    
    # For each edge (u,v), find common neighbors
    for u, v in edges:
        # Find nodes that are neighbors of both u and v
        common = graph[u] & graph[v]
        
        # For each common neighbor, form a triangle
        for w in common:
            # Sort to ensure unique representation
            triangle = tuple(sorted([u, v, w]))
            triangles.add(triangle)
    
    # Filter triangles containing 't' nodes
    return sum(1 for t in triangles if any(node.startswith('t') for node in t))

def test():
    assert find_triangles(sample_input) == 7
    print("Test passed!")

test()
```

Would you like me to explain any part of this implementation, or should we try it with the full input?

In [ ]:
from collections import defaultdict

def build_graph(edges):
    # Create adjacency list using defaultdict
    graph = defaultdict(set)
    for u, v in edges:
        graph[u].add(v)
        graph[v].add(u)
    return graph

def find_triangles(edges):
    graph = build_graph(edges)
    triangles = set()
    
    # For each edge (u,v), find common neighbors
    for u, v in edges:
        # Find nodes that are neighbors of both u and v
        common = graph[u] & graph[v]
        
        # For each common neighbor, form a triangle
        for w in common:
            # Sort to ensure unique representation
            triangle = tuple(sorted([u, v, w]))
            triangles.add(triangle)
    
    # Filter triangles containing 't' nodes
    return sum(1 for t in triangles if any(node.startswith('t') for node in t))

def test():
    assert find_triangles(sample_input) == 7
    print("Test passed!")

test()

Test passed!


In [ ]:
def solve_part1():
    return find_triangles(full_input)

solve_part1()

1240

that was correct answer

### Part 2

--- Part Two ---

There are still way too many results to go through them all. You'll have to find the LAN party another way and go there yourself.

Since it doesn't seem like any employees are around, you figure they must all be at the LAN party. If that's true, the LAN party will be the largest set of computers that are all connected to each other. That is, for each computer at the LAN party, that computer will have a connection to every other computer at the LAN party.

In the above example, the largest set of computers that are all connected to each other is made up of co, de, ka, and ta. Each computer in this set has a connection to every other computer in the set:

```
ka-co
ta-co
de-co
ta-ka
de-ta
ka-de
```

The LAN party posters say that the password to get into the LAN party is the name of every computer at the LAN party, sorted alphabetically, then joined together with commas. (The people running the LAN party are clearly a bunch of nerds.) In this example, the password would be co,de,ka,ta.

What is the password to get into the LAN party?

**AI Prompt**

what algorithms are known for finding maximum clique?

**AI Response**

Finding the maximum clique is NP-hard, but there are several well-known algorithms:

1. Bron-Kerbosch algorithm:
   - Most popular for practical use
   - Recursive backtracking with pivot selection
   - Works well for sparse graphs
   - Has variants with pruning optimizations

2. Branch and Bound algorithms:
   - Uses upper/lower bounds to prune search space
   - Can be faster for dense graphs
   - More complex implementation

3. Greedy approximation:
   - Much faster but not guaranteed to find maximum
   - Usually used for very large graphs
   - Can be used to get initial bound for exact algorithms

Given our graph size (520 nodes, 3380 edges) and that it's relatively sparse, would you like to explore implementing Bron-Kerbosch? It's typically the go-to choice for problems of this scale.

**AI Prompt**

lets use Bron-Kerbosch, please explain me how it works

**AI Response**

Let me explain Bron-Kerbosch algorithm step by step:

The algorithm maintains three sets:
1. R: the current clique being built
2. P: prospective vertices that could extend the clique
3. X: vertices already processed (used to avoid duplicates)

Basic algorithm flow:
1. Start with R empty, P containing all vertices, X empty
2. If both P and X are empty, R is a maximal clique - report it
3. For each vertex v in P:
   - Add v to R
   - Create new P' from vertices in P connected to v
   - Create new X' from vertices in X connected to v
   - Recurse with (R ∪ {v}, P', X')
   - Move v from P to X

The pivoting variant (more efficient) adds:
- Choose a pivot vertex u from P ∪ X
- Only process vertices in P that aren't neighbors of u

Would you like me to show a small example of how this works on a tiny graph, or should we proceed with implementation?

**AI Prompt**

show on tiny graph

**AI Response**

Let's use a tiny graph with 4 vertices that form a clique (complete subgraph) plus one extra vertex:

```
A --- B
|\   /|
| \ / |
|  C  |
| / \ |
|/   \|
D --- E
```

Let's walk through it step by step:

Initial state:
- R = {}  (current clique)
- P = {A,B,C,D,E}  (candidates)
- X = {}  (processed)

Should I continue with the first few steps of how the algorithm processes this graph? We can follow what happens when we start with vertex A.

**AI Prompt**

continue

**AI Response**

Starting with vertex A:

Step 1: Choose A
- R = {A}  (add A to current clique)
- P = {B,C,D}  (only A's neighbors remain as candidates)
- X = {}  (nothing processed yet)

Step 2: Choose B from P
- R = {A,B}
- P = {C}  (only common neighbors of A and B)
- X = {}

Step 3: Choose C from P
- R = {A,B,C}
- P = {}  (no vertices connected to all of A,B,C)
- X = {}

At this point, we've found a maximal clique {A,B,C}. The algorithm will backtrack and try other paths.

Would you like me to continue with what happens when we backtrack, or is this clear so far?

**AI Prompt**

continue

**AI Response**

Okay, let's see what happens when we backtrack:

After finding {A,B,C}, we backtrack from C:
- Move C to X
- R = {A,B}
- P = {}  (no more candidates with B)
- X = {C}  (C is processed)

Backtrack from B:
- Move B to X
- R = {A}
- P = {D}  (D is still a candidate from A's neighbors)
- X = {B,C}  (B and C processed)

Try path with D:
- R = {A,D}
- P = {}  (no vertices connected to both A and D that we haven't tried)
- X = {B,C}

This finds another maximal clique {A,D}.

Should I continue with what happens when we backtrack from A completely and start with a new vertex?

**AI Prompt**

continue

**AI Response**

Backtrack completely from A:
- Move A to X
- R = {}
- P = {B,C,D,E}  (E wasn't connected to A, so now available)
- X = {A}  (A is processed)

Start with B:
- R = {B}
- P = {C,E}  (B's unprocessed neighbors)
- X = {A}  (A was processed)

Try path with C:
- R = {B,C}
- P = {}  (no unprocessed common neighbors)
- X = {A}

This finds maximal clique {B,C}.

The algorithm continues this pattern, eventually finding all maximal cliques. In this example, the largest ones are size 3 (like {A,B,C}).

Would you like to see how the pivoting optimization would work on this same example, or should we move to implementing the algorithm?

**AI Prompt**

I understand, lets move to implementation

**AI Response**

I'll help you implement the Bron-Kerbosch algorithm. Let's start with the basic version first, then we can add pivoting if needed.

Here's a proposed structure:

```python
def find_max_clique(edges):
    # Build graph as adjacency sets
    graph = build_graph(edges)  # we already have this from part 1
    
    def bron_kerbosch(R, P, X):
        if not P and not X:
            return R
        
        max_clique = set()
        for v in set(P):  # make copy since we'll modify P
            neighbors = graph[v]
            new_r = R | {v}
            new_p = P & neighbors
            new_x = X & neighbors
            clique = bron_kerbosch(new_r, new_p, new_x)
            if len(clique) > len(max_clique):
                max_clique = clique
            P.remove(v)
            X.add(v)
        return max_clique

    return bron_kerbosch(set(), set(graph.keys()), set())

def test_max_clique():
    test_input = """
    A-B
    A-C
    A-D
    B-C
    C-D
    B-E
    C-E
    D-E
    """.strip().split('\n')
    edges = [edge.strip().split('-') for edge in test_input]
    clique = find_max_clique(edges)
    print(f"Found max clique: {clique}")
    assert len(clique) == 3  # should find a clique of size 3

test_max_clique()
```

Would you like me to explain any part of this implementation before we try it with our puzzle input?

In [ ]:
def find_max_clique(edges):
    # Build graph as adjacency sets
    graph = build_graph(edges)  # we already have this from part 1
    
    def bron_kerbosch(R, P, X):
        if not P and not X:
            return R
        
        max_clique = set()
        for v in set(P):  # make copy since we'll modify P
            neighbors = graph[v]
            new_r = R | {v}
            new_p = P & neighbors
            new_x = X & neighbors
            clique = bron_kerbosch(new_r, new_p, new_x)
            if len(clique) > len(max_clique):
                max_clique = clique
            P.remove(v)
            X.add(v)
        return max_clique

    return bron_kerbosch(set(), set(graph.keys()), set())

def test_max_clique():
    clique = find_max_clique(sample_input)
    print(f"Found max clique: {','.join(sorted(clique))}")
    assert len(clique) == 4  # should find a clique of size 3

test_max_clique()

Found max clique: co,de,ka,ta


In [ ]:
def solve_part2():
    clique = find_max_clique(full_input)
    print(f"Found max clique: {','.join(sorted(clique))}")

solve_part2()

Found max clique: am,aq,by,ge,gf,ie,mr,mt,rw,sn,te,yi,zb
